cips/cip-tx-interface.md

```diff
@@ -0,0 +1,90 @@
+
+Currently both these things can be provided if you have access to a trusted full node (which can simulate the transaction and return the global min gas price - which is sufficient in non-congested periods). This could be improved on in a few dimensions:
+
+- Estimating the gas by simulating the execution of a transaction is a heavy procedure
+- The minimum gas price is insufficient when there is congestion
+- Not all nodes expose these endpoints and it may not be so simple to find endpoints that are trusted
+
+In addition, Keplr provides some form of gas estimation but no gas price estimation. Users have to pick from "low", "medium", or "high"
+
+## Specification
+
+The following API is proposed for the standardised gas estimation service.
+
+```proto
+service GasEstimator {
+    rpc EstimateGasUsage(EstimateGasUsageRequest) returns (EstimateGasUsageResponse) {}
+    rpc EstimateGasPrice(EstimateGasPriceRequest) returns (EstimateGasPriceResponse) {}
+}
+
+message EstimateGasUsageRequest {
+    cosmos.tx.Tx tx = 1;
+}
+
+message EstimateGasUsageResponse {
+    uint64 estimated_gas_used = 1;
+}
+
+message EstimateGasPriceRequest {}
+
+message EstimateGasPriceResponse {
+    double estimated_gas_price = 1;
+}
+```
+
+Given it's wide usage both in the Cosmos SDK and more broadly, the service would be implemented using gRPC. RESTful endpoints may optionally be supported in the future using something like `grpc-gateway`.
+
+Given the expected reliance on clients, all future changes must be done in a strictly non-breaking way.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+## Rationale
+
+Most of the rationale behind the necessity of a standard interface and it's definition is captured in the motivation section.
+
+The notion of urgency, or within how many blocks does the user want their transaction to be submitted may be a useful parameter in the future but is currently left out of the original interface.
+
+## Backwards Compatibility
+
+As this is a new interface, no consideration for backwards compatibility is necessary
+
+## Reference Implementation
+
+The effectiveness of a "standardised" interface depends on the willingness of current and future clients to adopt it as well as the willingness of teams to provide those services. To set a sufficient precendent, both the Node API within `celestia-node` and the consensus node within `celestia-app` will implement client and server implementations respectively, creating an interface between the existing line of communication. That way by default, light nodes will use that API with the trusted provider they are already using for transaction submission.
+
+The consensus node will use the SimulateTx method to estimate the gas used and use the `min_gas_price` parameter within state as the `estimated_gas_price`
```

Comment:
I think `SimulateTx` is often used in conjunction with 

```
      --gas-adjustment float     adjustment factor to be multiplied against the estimate returned by the tx simulation; if the gas limit is set manually this flag is ignored  (default 1)
```

so should the initial implementation return the direct output of `SimulateTx` or multiplied by some factor (i.e. `1.1` or `1.2`)?

Reply:
I don't have strong feelings on this. I would lean towards multiplying by 1.1